### PR TITLE
Dockerfile improvements

### PR DESCRIPTION
**Pull request by ricardogsilva**
# Overview

This PR improves the existing `Dockerfile` by adding a few missing stuff from the first implementation. Namely:

- Addition of a `--reload` flag that makes the web server automatically reload whenever the code changes. This is only useful for development purposes, as it simplifies the dev workflow;
- Usage of python's `os.execlp` function, which behaves similarly to bash's `exec`. This causes the entrypoint script to be replaced by the web server (in this case gunicorn) in such a way that the same environment and PID is used. This effectively means that gunicorn is now PID1 of the docker container, making it easier to interact with it. It also means that a running pycsw container will now exit cleanly without errors when it is stopped or removed.

# Related Issue / Discussion
#530  
#534

# Additional Information

I have another also PR #547, which is related to this one. It adds documentation on how to use the docker image.

# Contributions and Licensing

(as per https://github.com/geopython/pycsw/blob/master/CONTRIBUTING.rst#contributions-and-licensing)

- [x] I'd like to contribute [feature X|bugfix Y|docs|something else] to pycsw. I confirm that my contributions to pycsw will be compatible with the pycsw license guidelines at the time of contribution.
- [x] I have already previously agreed to the pycsw Contributions and Licensing Guidelines

### CODE REVIEW

**Comment by ricardogsilva** [comment]
It seems I've messed up this PR with the documentation one #547. Sorry about that. 
Let me try to clean things up.

**Comment by ricardogsilva** [comment]
OK, all good now. Please review at your convenience.

**Comment by tomkralidis** [comment]
+1

File: requirements-dev.txt [hunk 1]
```diff
@@ -8,5 +8,4 @@ pytest==3.0.3
 pytest-cov==2.4.0
 pytest-flake8==0.8.1
 pytest-timeout==1.2.0
-requests==2.10.0
```
Comment (kalxas) [inline requirements-dev.txt]: is this removal related to this PR?

Comment (ricardogsilva) [reply requirements-dev.txt]: Actually (and strangely) yes it is.

The `requests` dependency does not seem to be needed for pycsw - I've done some grepp'ing in the codebase and I don't think we are using it anywhere (I might be wrong though).
 
In addition to that (and hence why it is related here), the gunicorn wsgi server, which is being used in the docker image, depends on a more recent version of requests. 
When I was testing the workflow for getting a development environment based on docker (as described in the docs on the other PR) I realized that when one would do 

```
pip install -r requirements-dev.txt
```

inside a running container it would uninstall the version of requests required by gunicorn and downgrade it to the one pinned in our requirements-dev.txt. This would lead to a crash in gunicorn when reloading/restarting the server's worker processes.

Comment (kalxas) [reply requirements-dev.txt]: thanks!
